datasync/chaindatafetcher/kas/repository_transactions.go

```diff
@@ -0,0 +1,181 @@
+
+	types.TxTypeValueTransfer:                      true,
+	types.TxTypeFeeDelegatedValueTransfer:          true,
+	types.TxTypeFeeDelegatedValueTransferWithRatio: true,
+
+	types.TxTypeValueTransferMemo:                      true,
+	types.TxTypeFeeDelegatedValueTransferMemo:          true,
+	types.TxTypeFeeDelegatedValueTransferMemoWithRatio: true,
+
+	types.TxTypeSmartContractDeploy:                      true,
+	types.TxTypeFeeDelegatedSmartContractDeploy:          true,
+	types.TxTypeFeeDelegatedSmartContractDeployWithRatio: true,
+
+	types.TxTypeSmartContractExecution:                      true,
+	types.TxTypeFeeDelegatedSmartContractExecution:          true,
+	types.TxTypeFeeDelegatedSmartContractExecutionWithRatio: true,
+}
+
+// transformToTxs transforms a chain event to transaction list according to the KAS database scheme.
+func transformToTxs(event blockchain.ChainEvent) []*Tx {
+	var txs []*Tx
+	// TODO-ChainDataFetcher extract EOAs in order to expire the KAS cache
+	//expiredEOAs := make(map[common.Address]struct{})
+
+	block := event.Block
+	head := block.Header()
+	receipts := event.Receipts
+
+	for idx, rawTx := range block.Transactions() {
+		txId := head.Number.Int64()*maxTransactionCount*maxTransactionLogCount + int64(idx)*maxInternalTransactionCount
+
+		// from
+		var from common.Address
+		if rawTx.IsLegacyTransaction() {
+			signer := types.NewEIP155Signer(rawTx.ChainId())
+			from, _ = types.Sender(signer, rawTx)
+		} else {
+			from, _ = rawTx.From()
+		}
+
+		// to
+		to := rawTx.To()
+		if to == nil {
+			to = &common.Address{}
+		}
+
+		// value
+		value := hexutil.EncodeBig(rawTx.Value())
+
+		// status
+		var status int
+		if receipts[idx].Status != types.ReceiptStatusSuccessful {
+			status = int(types.ReceiptStatusFailed)
+		} else {
+			status = int(receipts[idx].Status)
+		}
+
+		// fee payer
+		// fee ratio
+		var (
+			feePayer []byte
+			feeRatio uint
+		)
+
+		if rawTx.IsFeeDelegatedTransaction() {
+			payer, _ := rawTx.FeePayer()
+			feePayer = payer.Bytes()
+		}
+
+		ratio, ok := rawTx.FeeRatio()
+		if ok {
+			feeRatio = uint(ratio)
+		}
+		// TODO-ChainDataFetcher extract EOAs in order to expire the KAS cache
+		//expiredEOAs[from] = struct{}{}
+		//expiredEOAs[to] = struct{}{}
+
+		tx := &Tx{
+			Timestamp:       head.Time.Int64(),
+			TransactionId:   txId,
+			FromAddr:        from.Bytes(),
+			ToAddr:          to.Bytes(),
+			Value:           value,
+			TransactionHash: rawTx.Hash().Bytes(),
+			Status:          status,
+			TypeInt:         int(rawTx.Type()),
+			GasPrice:        rawTx.GasPrice().Uint64(),
+			GasUsed:         receipts[idx].GasUsed,
+			FeePayer:        feePayer,
+			FeeRatio:        feeRatio,
+		}
+
+		txs = append(txs, tx)
+	}
+	return txs
+}
+
+// InsertTransactions inserts transactions in the given chain event into KAS database.
+func (r *repository) InsertTransactions(event blockchain.ChainEvent) error {
+	return r.insertTransactions(transformToTxs(event))
```

Comment:
It seems that `insertTransactions` can be directly used here, doesn't it?

Reply:
Right. Applied on 8b05945844acaebf5f4581850db0bf10f66460bf.